### PR TITLE
Testing My_enumerable

### PR DESCRIPTION
**Pull request by Tahirbhalli**
Ruby
Rspec
each enumerable method is tested with positive and negative cases
test the enumerable with originals like my_each==each

### CODE REVIEW

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @Tahirbhalli and @sddoe :balloon: :lollipop: 

### This is a good start, :medal_sports:  we just need to add a few more cases and organize them better please keep reading :point_down: 

## Project Specific

- [ ] All methods implemented in the Enumerable project are tested ([screensho](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/all_enumerables.png)t from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- The Odin project only requires test cases for six methods but in the Microverse specifications is requested to add methods for all of the methods in your Enumerable methods project so please build at least one test case for each of your Enumerable methods project.

* [ ] specs should cover positive and negative scenarios ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/commonsense.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/)) 

* [ ] Specs for each method should be wrapped in one `describe` block. As in example here https://github.com/rubocop-hq/rspec-style-guide#empty-lines-around-examples . ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/describe.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- [ ] You should have one `it` block for each expected behaviour. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/itblock.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

* [ ] Instead of using `before_each` to define a variable for the test cases, you should [Use `let` instead](https://github.com/rubocop-hq/rspec-style-guide#let-blocks). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/before_each.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))

- here is an example of how you should properly create your positive and negative (using not_to) test cases defining your variables using let (I'm using your my_any), please do it like this for all 9 methods in your Enumerable methods project

```
RSpec.describe Enumerable do
  let(:arr) { [1, 2, 3, 4, 5] }
  describe '#my_any?' do
    it 'expects to retrieve the same result as the real any? method' do
      expect(arr.my_any?).to eq(arr.any?)
    end
    it 'if no block is given returns true if at least one element of the enum is truthy' do
      expect(arr.my_any?).not_to eq(false)
    end
 end
.
.
end

```

## General

- [x] Pull Request has meaningful title and a short description ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_prs.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

- Please edit your pull request and add a couple of lines more to describe what you accomplish by building this project.

- [x] Commits are descriptive ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_commits.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

- Commit more often (maybe every time you finish a test case) so it's easier to describe and each commit doesn't have to be that general, also I don't see any commits from Sergio Perdomo, I don't even see Sergio Perdomo as a collaborator, remember this is a collaborative project so there should be proportional commits from both students.

### That's it, team! :raised_hands: 

### Please :pray: work on these changes :man_technologist: :man_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any question? please ask me on slack **@ Maru**

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Hey Guys
I appreciate the effort you have put into this project but we still need to do more before we can close this off.
The previous reviewer suggested a lot of crucial things that you guys failed to implement.
### [REQUIRED]
- [x] Test all methods in the enumerable methods project. There are about 9 methods so have about 9 describe blocks
```ruby
RSpec.describe Enumerable do
 let(:arr) { [1, 2, 3, 4, 5] }
 describe '#my_any?' do
    it 'expects to retrieve the same result as the real any? method' do
      expect(arr.my_any?).to eq(arr.any?)
    end
    it 'if no block is given returns true if at least one element of the enum is truthy' do
      expect(arr.my_any?).not_to eq(false)
    end
 end
 describe '#my_all?' do
    it 'expects to retrieve the same result as the real all? method' do
      expect(arr.my_all?).to eq(arr.all?)
    end
 end
end
```
- [x] Test both positive and negative scenarios
- [x] Resubmit when done with the changes

### Martin Afani

**Review by pelzolga123 — CHANGES_REQUESTED** [review]
Project requires changes.

**Comment by pelzolga123** [comment]
## Hi Team!  
### Good job so far :+1: 
Still you haven't done everything required by previous revievers:
* [x] Not all methods have positive and negative test cases.
* [x] You have only one `describe` block, that is for `Enumerable`. You need to create 9 more `describe` blocks for each method you are testing.

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

File: spec/enum_spec.rb [hunk 1]
```diff
@@ -1,74 +1,108 @@
...
+      expect([].all?.equal?([].my_all?)).to eq(true)
+    end
+    it 'my_all negative case' do
+      expect([1, 2].all?.equal?([2, 1].my_all?)).not_to eq(false)
+    end
+    it 'my_all' do
+      words = %w[bacon orange apple]
+      expect(words.my_all?).to eq(words.all?)
+    end
+  end
```
Comment (bolah2009) [inline spec/enum_spec.rb:43]: Testing positive and negative cases means the test covers all the condition branches. 

The test still passes with the following commented out code.

```ruby
  def my_all?(proc = nil)
    ar = self
    return ar.all?(proc) if proc

    if block_given?
      ar.my_each do |i|
        next unless (yield i) == false

        return false
      end
      # return true
    end

    return false if ar.include? false or ar.include? nil

    true
  end
```

File: spec/enum_spec.rb [hunk 2]
```diff
@@ -1,74 +1,108 @@
...
+      words = %w[bacon orange apple]
+      expect(words.none? { |str| str.size == 5 } == words.my_none? { |str| str.size == 5 }).to eq(true)
+    end
+    it 'my_none poitive case' do
+      expect([].none?.equal?([].my_none?)).to eq(true)
+    end
+    it 'my_none negative case' do
+      expect([1, 2].none?.equal?([1, 2].my_none?)).not_to eq(false)
+    end
+  end
```
Comment (bolah2009) [inline spec/enum_spec.rb:55]: [Optional]

Like the previous comment, the commented out code below till passes the test


```ruby
  def my_none?(proc = nil)
    ar = self
    return ar.none?(proc) if proc

    if block_given?
      ar.my_each do |i|
        next unless (yield i) == true

        return false
      end
      # return true
    end
    !ar[ar.length - 1]
  end
```

File: spec/enum_spec.rb [hunk 3]
```diff
@@ -1,74 +1,108 @@
...
+    end
+    it 'my_any negative' do
+      arr = [1, 2, 3, 4, 5]
+      expect(arr.my_any?.equal?(arr.any?)).not_to eq(false)
+    end
+    it 'my_any with block' do
+      arr = [1, 2, 3, 4, 5]
+      expect(arr.my_any? { |i| i < 3 }).to eq(arr.any? { |i| i < 3 })
+    end
+  end
```
Comment (bolah2009) [inline spec/enum_spec.rb:27]: [Optional]

Like the previous comment, the commented out code below till passes the test


```ruby
  def my_any?(proc = nil)
    ar = self
    return ar.any?(proc) if proc

    if block_given?
      ar.my_each do |i|
        next unless (yield i) == true

        return true
      end
      # return false
    end
    return true if !ar.member? nil or ar.include? true

    # false
  end
```

File: spec/enum_spec.rb [hunk 4]
```diff
@@ -1,74 +1,108 @@
...
+    end
+    it 'my_count positive case' do
+      arr = [1, 2, 3, 4, 5]
+      expect(arr.my_count { |i| i < 3 }).to eq(arr.count { |i| i < 3 })
+    end
+    it 'my_count negative case' do
+      arr = [1, 2, 3, 4, 5]
+      expect(arr.my_count.equal?(4)).to eq(false)
+    end
+  end
```
Comment (bolah2009) [inline spec/enum_spec.rb:81]: [Optional]

Like the previous comment, the commented out code below till passes the test


```ruby
  def my_count(proc = nil)
    ar = self
    i = 0
    # if proc
    #   ar.my_each do |x|
    #     next unless x == proc

    #     i += 1
    #   end
    #   return i
    # end
    if block_given?
      ar.my_each do |x|
        next unless (yield x) == true

        i += 1
      end
      return i
    end

    ar.my_each do
      i += 1
    end
    i
  end
```

File: spec/enum_spec.rb [hunk 5]
```diff
@@ -1,74 +1,108 @@
...
+  describe 'map' do
+    it 'my_map' do
+      arr = [1, 2, 3, 4, 5]
+      expect(arr.my_map { |x| x * 2 } == arr.map { |x| x * 2 }).to eq(true)
+    end
+    it 'my_map' do
+      arr = [1, 2, 3, 4, 5]
+      expect(arr.my_map.class.equal?(arr.map.class)).to eq(true)
+    end
   end
```
Comment (bolah2009) [inline spec/enum_spec.rb]: [Optional]

Like the previous comment, the commented out code below till passes the test


```ruby
  def my_map(proc = nil)
    ar = self
    new_arr = []
    # unless proc.nil?
    #   ar.each do |x|
    #     return false unless proc.yield x
    #   end
    #   return true
    # end
    return to_enum unless block_given? && proc.nil?

    ar.my_each do |j|
      a = yield j
      new_arr.append(a)
    end
    new_arr
  end
```

**Review by bolah2009 — APPROVED** [review]
Good job with all the requested changes from previous reviewers, It's a 👍🏾 from me!

### :heavy_check_mark: Approved!

Check out the optional comments before merging. :relaxed: 

Good luck with your next projects. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

**Pull request merged**